### PR TITLE
improve end condition for currentEventsByTag, #206

### PR DESCRIPTION
**Pull request by patriknw**
Refs #206

### CODE REVIEW

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala [hunk 1]
```diff
@@ -851,3 +851,42 @@ class EventsByTagStrictBySeqNoEarlyFirstOffsetSpec
   }
 }
 
+class EventsByTagStrictBySeqNoManyInCurrentTimeBucketSpec
```
Comment (patriknw) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: @wketting can you try this test and enable the `Thread.sleep(30)` in the EventsByTagFetcher

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagFetcher.scala [hunk 2]
```diff
@@ -113,6 +113,9 @@ private[query] class EventsByTagFetcher(
           else
             highestOffset = offs
 
+          // FIXME reproducer of issue #206
+          // Thread.sleep(30)
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagFetcher.scala]: enable this for reproducer

Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagFetcher.scala]: the problem is when replaying the last time bucket takes long time, so that the `LookingForDelayed` mode is triggered (over and over again)

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala [hunk 3]
```diff
@@ -263,12 +263,12 @@ private[query] class EventsByTagPublisher(
...
       if (log.isDebugEnabled)
         log.debug(
-          s"${backtrackingLog}query chunk done for tag [{}], timBucket [{}], count [{}] of [{}]",
+          s"${backtrackingLog}query chunk done for tag [{}], timBucket [{}], count [{}] of [{}], mightBeMore [$mightBeMore]",
           tag, currTimeBucket, deliveredCount, retrievedCount
         )
       seqNumbers = seqN
       currOffset = highest
-      replayCountZero = retrievedCount == 0
+      replayCountZero = backtrackingMode == NoBacktracking && retrievedCount == 0
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: `replayCountZero` is used in the end condition for `currentEventsByTag`.
I wonder if it makes sense to use the `LookingForDelayed` for the `currentEventsByTag` at all. The purpose is to find new persistence ids while the query is running and that is good for the "live" query but not sure if it makes any sense for the "current" query.

Comment (wketting) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: This thought crossed my mind as well. Imho LookingForDelayed is not needed for currentEventsByTag.

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala [hunk 4]
```diff
@@ -280,6 +280,9 @@ private[query] class EventsByTagPublisher(
       } else {
         if (backtrackingMode != NoBacktracking)
           currTimeBucket = TimeBucket(currOffset)
+        // make sure LookingForDelayed is not triggered immediately again in case of slow queries
+        if (backtrackingMode == LookingForDelayed)
+          lookForDelayedDeadline = nextLookForDelayedDeadline()
         backtrackingMode = NoBacktracking
         stopIfDone()
       }
```
Comment (wketting) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: This solves the issue.

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala (hunk 1 above)
Comment (wketting) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: I've tested this a view times, also locally in our test environment. The bug is not occurring anymore. So the solution is looking good! I still find it very difficult to determine the exact flow in this matter (probably also due to my lack of knowledge of the inner workings of the library), but the test is providing similar logs (when failing) as I had on my test env. The fix provided here does make sense imho.

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala (hunk 3 above)
Comment (wketting) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: I'm not sure why this change would help. I thought initially that replayCountZero being false (due to duplicate event found) caused lookingfordelayed to loop.

Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: We want to reach the end condition (`replayCountZero == true`) by a normal query not via a LookingForDelayed query.
However, we should disable the LookingForDelayed altogether for currentEventsByTag.

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala (hunk 1 above)
Comment (patriknw) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: thanks for testing

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala (hunk 3 above)
Comment (wketting) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: Aha, clear (and agree).

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala (hunk 1 above)
Comment (wketting) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: @patriknw Unfortunately after some more testing today, I ran again into the same (hanging loop) situation as earlier. Somehow I cannot recreate the situation in a test. Although I think a potential bug is solved with this PR, there still is an issue. It would be really helpful to have an option to switch off lookingForDelayed during replay.

Comment (patriknw) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: thanks for more testing, I have pushed another commit to this PR, disabling the LookingForDelayed for currentEventsByTag

Comment (wketting) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: I will keep on testing to coming days.

Comment (wketting) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: @patriknw We've tested and used this fix for some days now. Our original issue has not occurred anymore. I would appreciate it if this PR could become a release soon. Would that be possible?

Comment (patriknw) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: thanks, then I'll release today

File: core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala [hunk 5]
```diff
@@ -80,7 +80,7 @@ object EventsByTagSpec {
     """).withFallback(CassandraLifecycle.config)
 
   val strictConfig = ConfigFactory.parseString(s"""
-    akka.loglevel = INFO
+    akka.loglevel = DEBUG
```
Comment (johanandren) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: Left at `DEBUG` by mistake?

Comment (johanandren) [inline core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala]: Left at `DEBUG` by mistake?

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagFetcher.scala (hunk 2 above)
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagFetcher.scala]: Leave comment for documentation purposes, but it's not really a `FIXME` so remove that?

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala [hunk 6]
```diff
@@ -199,8 +199,10 @@ private[query] class EventsByTagPublisher(
         context.system.scheduler.scheduleOnce(settings.eventualConsistencyDelay, self, Continue)
   }
 
-  def timeToLookForDelayed: Boolean =
-    strictBySeqNumber && !backtracking && lookForDelayedDeadline.isOverdue() && !isTimeBucketBeforeToday()
+  def timeToLookForDelayed: Boolean = {
+    strictBySeqNumber && toOffset.isEmpty && !backtracking &&
+      lookForDelayedDeadline.isOverdue() && !isTimeBucketBeforeToday()
+  }
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: Hm, so if it is a `currentEventsByTag` `toOffset` will contain the end timestamp and this always return `false? Not sure how this relates to the other changes.

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala [hunk 7]
```diff
@@ -280,6 +282,9 @@ private[query] class EventsByTagPublisher(
       } else {
         if (backtrackingMode != NoBacktracking)
           currTimeBucket = TimeBucket(currOffset)
+        // make sure LookingForDelayed is not triggered immediately again in case of slow queries
+        if (backtrackingMode == LookingForDelayed)
+          lookForDelayedDeadline = nextLookForDelayedDeadline()
```
Comment (johanandren) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: I wish we could refactor this actor to be less `if a ..., if b ..., if else if if`.  Maybe outside the scope of this PR but I find it makes me need to keep more things in my head than I want to when reading these sources.

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagFetcher.scala (hunk 2 above)
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagFetcher.scala]: I will remove this, and DEBUG before merging

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala (hunk 7 above)
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: yes, should be part of the GraphStage refactoring

File: core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala (hunk 6 above)
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/query/EventsByTagPublisher.scala]: this would probably be the only thing need to fix the reported issue with `currentEventsByTag`, but I wanted to understand the root cause and fix that also since that could be used in the live query (so the issue is fixed kind of twice)
